drv/sidecar-mainboard-controller/src/ignition.rs

```diff
@@ -0,0 +1,143 @@
+        Self {
+            fpga: FpgaUserDesign::new(
+                task_port,
+                MainboardController::DEVICE_INDEX,
+            ),
+        }
+    }
+
+    #[inline]
+    fn port_addr(&self, port: u8, offset: IgnitionAddr) -> u16 {
+        u16::from(Addr::IGNITION_CONTROLLERS_COUNT)
+            + 0x100
+            + (0x100 * port as u16)
+            + u16::from(offset)
+    }
+
+    fn read_port_register<T>(
+        &self,
+        port: u8,
+        offset: IgnitionAddr,
+    ) -> Result<T, FpgaError>
+    where
+        T: AsBytes + Default + FromBytes,
+    {
+        self.fpga.read(self.port_addr(port, offset))
+    }
+
+    fn write_port_register<T>(
+        &self,
+        port: u8,
+        offset: IgnitionAddr,
+        value: T,
+    ) -> Result<(), FpgaError>
+    where
+        T: AsBytes + Default + FromBytes,
+    {
+        self.fpga
+            .write(WriteOp::Write, self.port_addr(port, offset), value)
+    }
+
+    /// Return the number of ports exposed by the Controller.
+    pub fn port_count(&self) -> Result<u8, FpgaError> {
+        self.fpga.read(Addr::IGNITION_CONTROLLERS_COUNT)
+    }
+
+    /// Return a bit-vector indicating Target presence on each of the Controller
+    /// ports.
+    pub fn presence_summary(&self) -> Result<u64, FpgaError> {
+        self.fpga.read(Addr::IGNITION_TARGETS_PRESENT0)
+    }
+
+    /// Return the state for the given port.
+    pub fn state(&self, port: u8) -> Result<PortState, FpgaError> {
+        self.read_port_register(port, IgnitionAddr::CONTROLLER_STATUS)
+            .map(PortState)
+    }
+
+    /// Return the high level counters for the given port.
+    pub fn counters(&self, port: u8) -> Result<Counters, FpgaError> {
+        self.read_port_register(
+            port,
+            IgnitionAddr::CONTROLLER_STATUS_RECEIVED_COUNT,
+        )
+    }
+
+    #[inline]
+    fn link_events_addr(link: LinkSelect) -> IgnitionAddr {
+        match link {
+            LinkSelect::Controller => {
+                IgnitionAddr::CONTROLLER_LINK_EVENTS_SUMMARY
+            }
+            LinkSelect::TargetLink0 => {
+                IgnitionAddr::TARGET_LINK0_EVENTS_SUMMARY
+            }
+            LinkSelect::TargetLink1 => {
+                IgnitionAddr::TARGET_LINK1_EVENTS_SUMMARY
+            }
+        }
+    }
+
+    /// Return the event summary vector for the given port and link.
+    pub fn link_events(
+        &self,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<LinkEvents, FpgaError> {
+        self.read_port_register(port, Self::link_events_addr(link))
+    }
+
+    /// Clear the events for the given port, link.
+    pub fn clear_link_events(
+        &self,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<(), FpgaError> {
+        self.write_port_register(
+            port,
+            Self::link_events_addr(link),
+            LinkEvents(0b111111), // Clear all events
```

Comment:
Could this be a constant on `LinkEvents`?

Reply:
Done, thanks!